core/src/apps/cardano/auxiliary_data.py

```diff
@@ -3,118 +3,171 @@
 
 if TYPE_CHECKING:
-    CatalystRegistrationPayload = dict[int, bytes | int]
-    SignedCatalystRegistrationPayload = tuple[CatalystRegistrationPayload, bytes]
-    CatalystRegistrationSignature = dict[int, bytes]
-    CatalystRegistration = dict[
-        int, CatalystRegistrationPayload | CatalystRegistrationSignature
+    Delegations = list[tuple[bytes, int]]
+    GovernanceRegistrationPayload = dict[int, Delegations | bytes | int]
+    SignedGovernanceRegistrationPayload = tuple[GovernanceRegistrationPayload, bytes]
+    GovernanceRegistrationSignature = dict[int, bytes]
+    GovernanceRegistration = dict[
+        int, GovernanceRegistrationPayload | GovernanceRegistrationSignature
     ]
 
     from . import seed
 
 AUXILIARY_DATA_HASH_SIZE = 32
-CATALYST_VOTING_PUBLIC_KEY_LENGTH = 32
-CATALYST_REGISTRATION_HASH_SIZE = 32
+GOVERNANCE_VOTING_PUBLIC_KEY_LENGTH = 32
+GOVERNANCE_REGISTRATION_HASH_SIZE = 32
 
-METADATA_KEY_CATALYST_REGISTRATION = 61284
-METADATA_KEY_CATALYST_REGISTRATION_SIGNATURE = 61285
+METADATA_KEY_GOVERNANCE_REGISTRATION = 61284
+METADATA_KEY_GOVERNANCE_REGISTRATION_SIGNATURE = 61285
+
+MAX_DELEGATION_COUNT = 32
+DEFAULT_VOTING_PURPOSE = 0
+
+
+def assert_cond(condition: bool) -> None:
+    if not condition:
+        raise wire.ProcessError("Invalid auxiliary data")
 
 
 def validate(auxiliary_data: messages.CardanoTxAuxiliaryData) -> None:
     fields_provided = 0
     if auxiliary_data.hash:
         fields_provided += 1
         _validate_hash(auxiliary_data.hash)
-    if auxiliary_data.catalyst_registration_parameters:
+    if auxiliary_data.governance_registration_parameters:
         fields_provided += 1
-        _validate_catalyst_registration_parameters(
-            auxiliary_data.catalyst_registration_parameters
+        _validate_governance_registration_parameters(
+            auxiliary_data.governance_registration_parameters
         )
-
-    if fields_provided != 1:
-        raise wire.ProcessError("Invalid auxiliary data")
+    assert_cond(fields_provided == 1)
 
 
 def _validate_hash(auxiliary_data_hash: bytes) -> None:
-    if len(auxiliary_data_hash) != AUXILIARY_DATA_HASH_SIZE:
-        raise wire.ProcessError("Invalid auxiliary data")
+    assert_cond(len(auxiliary_data_hash) == AUXILIARY_DATA_HASH_SIZE)
 
 
-def _validate_catalyst_registration_parameters(
-    catalyst_registration_parameters: messages.CardanoCatalystRegistrationParametersType,
+def _validate_governance_registration_parameters(
+    parameters: messages.CardanoGovernanceRegistrationParametersType,
 ) -> None:
-    if (
-        len(catalyst_registration_parameters.voting_public_key)
-        != CATALYST_VOTING_PUBLIC_KEY_LENGTH
-    ):
-        raise wire.ProcessError("Invalid auxiliary data")
+    voting_key_fields_provided = 0
+    if parameters.voting_public_key is not None:
+        voting_key_fields_provided += 1
+        _validate_voting_public_key(parameters.voting_public_key)
+    if parameters.delegations:
```

Comment:
What happens if the delegations are an empty array? According to the Slack discussions, IOG aims to interpret an empty delegation array as deregistration, so it will be a valid value (but AFAIK protobuf does not distinguish between no and empty array given?).

Reply:
No array should be the same as empty array in protobuf. This implementation still enforces the non-emptiness, let's see what will be the result of the discussion.

Reply:
No empty arrays allowed in a registration, thats the outcome so far in the discussion. As a compromise the usused key 61286 will be used to specify a deregistration metadata format. Its up to the tools/wallets to include such an option or not. This is the PR to add those specifications into CIP-0036: https://github.com/cardano-foundation/CIPs/pull/349